contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol

```diff
@@ -0,0 +1,710 @@
+        _withdrawFromGaugeAndPool(lpTokens);
+
+        // Burn all the removed OS and any that was left in the strategy
+        uint256 osToBurn = IERC20(os).balanceOf(address(this));
+        IVault(vaultAddress).burnForStrategy(osToBurn);
+
+        // Transfer wS to the recipient
+        // Note there can be a dust amount of wS left in the strategy as
+        // the burn of the pool's LP tokens is rounded up
+        require(
+            IERC20(ws).balanceOf(address(this)) >= _wsAmount,
+            "Not enough wS removed from pool"
+        );
+        IERC20(ws).safeTransfer(_recipient, _wsAmount);
+
+        // Ensure solvency of the vault
+        _solvencyAssert();
+
+        // Emit event for the withdrawn wS tokens
+        emit Withdrawal(_asset, pool, _wsAmount);
+        // Emit event for the burnt OS tokens
+        emit Withdrawal(os, pool, osToBurn);
+    }
+
+    /**
+     * @notice Withdraw all pool LP tokens from the gauge,
+     * remove all wS and OS from the SwapX pool,
+     * burn all the OS tokens,
+     * and transfer all the wS to the Vault contract.
+     * @dev There is no solvency check here as withdrawAll can be called to
+     * quickly secure assets to the Vault in emergencies.
+     */
+    function withdrawAll()
+        external
+        override
+        onlyVaultOrGovernor
+        nonReentrant
+        skimPool
+    {
+        // Get all the pool LP tokens the strategy has staked in the gauge
+        uint256 lpTokens = IGauge(gauge).balanceOf(address(this));
+        // Can not withdraw zero LP tokens from the gauge
+        if (lpTokens == 0) return;
+
+        // Withdraw pool LP tokens from the gauge and remove assets from from the pool
+        _withdrawFromGaugeAndPool(lpTokens);
+
+        // Burn all OS in this strategy contract
+        uint256 osToBurn = IERC20(os).balanceOf(address(this));
+        IVault(vaultAddress).burnForStrategy(osToBurn);
+
+        // Get the strategy contract's wS balance.
+        // This includes all that was removed from the SwapX pool and
+        // any that was sitting in the strategy contract before the removal.
+        uint256 wsBalance = IERC20(ws).balanceOf(address(this));
+        IERC20(ws).safeTransfer(vaultAddress, wsBalance);
+
+        // Emit event for the withdrawn wS tokens
+        emit Withdrawal(ws, pool, wsBalance);
+        // Emit event for the burnt OS tokens
+        emit Withdrawal(os, pool, osToBurn);
+    }
+
+    /***************************************
+                Pool Rebalancing
+    ****************************************/
+
+    /** @notice Used when there is more OS than wS in the pool.
+     * wS and OS is removed from the pool, the received wS is swapped for OS
+     * and the left over OS in the strategy is burnt.
+     * The OS/wS price is < 1.0 so OS is being bought at a discount.
+     * @param _wsAmount Amount of Wrapped S (wS) to swap into the pool.
+     */
+    function swapAssetsToPool(uint256 _wsAmount)
```

Comment:
the name could also be `swapWsToPool`

Reply:
I tried to keep the function name generic in case we reused this strategy for other assets or chains. 
The same was done for the Curve AMO which has `mintAndAddOTokens`, `removeAndBurnOTokens` and `removeOnlyAssets` rebalance functions. This worked out when the Sonic Curve AMO was done.

Reply:
makes sense I agree the same function names across multiple strategies is better.